### ISSUE TITLE
Issue with "inside_im" in visual model VGCPCFGs

### ISSUE THREAD

**Issue by evaportelance** [zhaoyanpeng/vpcfg#3]
Hi! In the file vpcfg/model_vis.py on line 96, `ll, span_margs = dist.inside_im`, in the following function:

```
def forward_parser(self, captions, lengths):
        params, kl = self.parser(captions)
        dist = SentCFG(params, lengths=lengths)
        the_spans = dist.argmax[-1]
        argmax_spans, trees, lprobs = utils.extract_parses(the_spans, lengths.tolist(), inc=0) 

        ll, span_margs = dist.inside_im
        nll = -ll
        kl = torch.zeros_like(nll) if kl is None else kl
        return nll, kl, span_margs, argmax_spans, trees, lprobs
```

the function inside_im is called on the dist SentCFG object. The problem is no such function or parameter exists for SentCFG in either the implementation of it in your forked modified version of torch_struct or the original version of torch_struct (I even looked back to commits as far as early 2019 and could not find this anywhere). This line is supposed to get the marginal probabilities for all possible spans for each item in batch from the inside-outside algorithm in order to be used for the image-span matching loss. However, since no function exists, it does not work. I am wondering if you made further modification to torch_struct that were never pushed or if you had additional source files that made your local environment definition of SentCFG() different from that of torch_struct (your repo or the original repo's implementation)? 

I did find that span marginals could be returned by calling dist.marginals[3], however they are of the format (batch_size, max_length, max_length, n_nt_rule) -- i.e. they represent spans as a matrix of  max_length x max_length, where impossible spans are zeros), when here span_margs is expected to be (batch_size, total_number_of_spans, n_nt_rules). It is not obvious how to transform (batch_size, max_length, max_length, n_nt_rule)  to (batch_size, total_number_of_spans, n_nt_rules), just removing zeros doesnt always give you the right number of spans for some reason...

Happy to discuss more!

**Comment by evaportelance** [comment]
Ah! I realise my error. You have many branches in your torch-struct repo and I had cloned the wrong one. issue solve.
